File: nodejs/exceljs/createExcelSheet.js
const odbc = require('odbc');
const Excel = require('exceljs');
const connectionString = require('./getConnectionString')();

async function createExcelSheet() {
  const connection = await odbc.connect(connectionString);
  const query = 'SELECT CUSNUM, LSTNAM, BALDUE, CDTLMT FROM QIWS.QCUSTCDT';
  const resultSet = await connection.query(query);

  const workbook = new Excel.Workbook();
  const worksheet = workbook.addWorksheet('Customers');

  worksheet.columns = [
    { header: 'id', key: 'CUSNUM', width: 10 },
    { header: 'Last Name', key: 'LSTNAM', width: 10 },
    { header: 'Balance Due', key: 'BALDUE', width: 11 },
    { header: 'Credit Limit', key: 'CDTLMT', width: 10 },
  ]

  for (const row of resultSet) {
    worksheet.addRow(row);
  }

  await workbook.xlsx.writeFile('Customers.xlsx');
}

createExcelSheet().catch((error) => {
  console.warn('Failed to create the excel worksheet');
  console.error(error);
});

File: python/xlsxwriter/create_excel_sheet.py
import pyodbc
import argparse
import os

from xlsxwriter import Workbook
from dotenv import load_dotenv

load_dotenv()

def get_connection_string():
    connection_string = "DSN=*LOCAL"

    if os.getenv("DSN"):
        connection_string = f"DSN={os.getenv('DSN')}"

    if os.getenv("DB_USER"):
        connection_string += f"DRIVER=IBM i Access ODBC Driver;UID={os.getenv('DB_USER')};"

        if not os.getenv("DB_PASS"):
            raise Exception("Database user was provided without a password")

        connection_string += f"PWD={os.getenv('DB_PASS')};SYSTEM={os.getenv('DB_HOST', 'localhost')};"

    return connection_string

conn = pyodbc.connect(get_connection_string())
cur = conn.cursor()
cur.execute("SELECT CUSNUM, LSTNAM, BALDUE, CDTLMT FROM QIWS.QCUSTCDT")

file_name = "Customers.xlsx"

with Workbook(file_name) as workbook:
    worksheet = workbook.add_worksheet("Customers")
    headers = [descr[0] for descr in cur.description]
    worksheet.write_row("A1", headers)

    for row, data in enumerate(cur, start=1):
        worksheet.write_row(row, 0, data)

print(f"Succssfully wrote {file_name}")

